File: src/find-the-index-of-the-first-occurrence-in-a-string.java

class Solution {
  public int strStr(String haystack, String needle) {
      if(needle.length() == 0 || needle == "" || needle.length() > haystack.length()) return -1;

      int needleLength = needle.length();
      for (int index = 0; index <= haystack.length() - needleLength; index++) {
        int j = 0;
        while(j < needleLength && haystack.charAt(index + j) == needle.charAt(j)) {
          j++;
        }
        if (j == needleLength) {
          return index;
        }
      }

      return -1;
  }
}

File: src/find-the-index-of-the-first-occurrence-in-a-string.js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 * https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/
 */
var strStr = function (haystack, needle) {
  let returnValue = -1;
  if (!needle || needle == "" || needle.length > haystack) return returnValue;
  if (haystack === needle) return 0;
  return isPart(haystack, needle);
};

function isPart(haystack, needle) {
  const needleLength = needle.length;
  for (let index = 0; index <= haystack.length - needleLength; index++) {
    let subString = haystack.substring(index, index + needleLength);
    if (subString.includes(needle)) {
      return index;
    }
  }

  return -1;
}
